src/main/scala/is/hail/methods/LinearMixedRegression.scala

```diff
@@ -283,236 +291,185 @@ class StandardLMM(lmmConstants: LMMConstants, delta: Double, logNullS2: Double,
+      val h2NormLkhd = IndexedSeq(Double.NaN) ++ h2Lkhd.map(_ / h2LkhdSum) ++ IndexedSeq(Double.NaN)
 
-        -0.5 * (logdetD + logdetCdC - logdetCtC + (n - d) * (math.log(sigma2))) + shift
-      }
+      (FastMath.exp(maxlogDelta), GlobalFitLMM(maxLogLkhd, gridLogLkhd, sigmaH2, h2NormLkhd))
     }
 
-    val minLogDelta = -8
-    val maxLogDelta = 8
-    val pointsPerUnit = 100 // number of points per unit of log space
-
-    val grid = (minLogDelta * pointsPerUnit to maxLogDelta * pointsPerUnit).map(_.toDouble / pointsPerUnit) // avoids rounding of (minLogDelta to logMax by logres)
-    val logLkhdFunction = if (useML) LogLkhdML else LogLkhdREML
-
-    val gridLogLkhd = grid.map(logDelta => (logDelta, logLkhdFunction.value(logDelta)))
-
-    val header = "logDelta\tlogLkhd"
-    val gridValsString = gridLogLkhd.map { case (d, nll) => s"$d\t$nll" }.mkString("\n")
-    log.info(s"\nlmmreg: table of delta\n$header\n$gridValsString\n")
-
-    val approxLogDelta = gridLogLkhd.maxBy(_._2)._1
-
-    if (approxLogDelta == minLogDelta)
-      fatal(s"lmmreg: failed to fit delta: ${if (useML) "ML" else "REML"} realized at delta lower search boundary e^$minLogDelta = ${FastMath.exp(minLogDelta)}, indicating negligible enviromental component of variance. The model is likely ill-specified.")
-    else if (approxLogDelta == maxLogDelta)
-      fatal(s"lmmreg: failed to fit delta: ${if (useML) "ML" else "REML"} realized at delta upper search boundary e^$maxLogDelta = ${FastMath.exp(maxLogDelta)}, indicating negligible genetic component of variance. Standard linear regression may be more appropriate.")
-
-    val searchInterval = new SearchInterval(minLogDelta, maxLogDelta, approxLogDelta)
-    val goal = GoalType.MAXIMIZE
-    val objectiveFunction = new UnivariateObjectiveFunction(logLkhdFunction)
-    val brentOptimizer = new BrentOptimizer(5e-8, 5e-7) // tol = 5e-8 * abs((ln(delta))) + 5e-7 <= 1e-6
-    val logDeltaPointValuePair = brentOptimizer.optimize(objectiveFunction, goal, searchInterval, MaxEval.unlimited)
-
-    val maxlogDelta = logDeltaPointValuePair.getPoint
-    val maxLogLkhd = logDeltaPointValuePair.getValue
-
-    if (math.abs(maxlogDelta - approxLogDelta) > 1d / pointsPerUnit) {
-      warn(s"lmmreg: the difference between the optimal value $approxLogDelta of ln(delta) on the grid and" +
-        s"the optimal value $maxlogDelta of ln(delta) by Brent's method exceeds the grid resolution" +
-        s"of ${1d / pointsPerUnit}. Plot the values over the full grid to investigate.")
-
+    val (delta, optGlobalFit) = optDelta match {
+      case Some(delta0) => (delta0, None)
+      case None => {
+        val (delta0, gf) = fitDelta()
+        (delta0, Some(gf))
+      }
     }
 
-    val epsilon = 1d / pointsPerUnit
-
-    // three values of ln(delta) right of, at, and left of the MLE
-    val z1 = maxlogDelta + epsilon
-    val z2 = maxlogDelta
-    val z3 = maxlogDelta - epsilon
+    def solve(): DiagLMM = {
```

Comment:
I'd change this to`def fitUsingDelta(delta: Double)` by contrast with `fitDelta()` and move it up above the block starting with `val (delta, optGlobalFit) = optDelta match {`.

Then end the class by running `fitUsingDelta(delta)`